Sync: improve importer detection and sync start action

jonathansadowski opened the pull request:
Take two on the work started in #11523. See D25761-code.

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Adds a utility function (`get_calling_importer_class`) which infers the import class responsible for calling the import action.
* Reworks the logic for triggering the sync action so that it is the same for `import_start`, `import_done`, and `import_end`.

#### Testing instructions:
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to `/wp-admin/admin.php?import=wordpress`
* Complete an import. If the import completes successfully, two actions should be synced
  * `jetpack_sync_import_start`
  * `jetpack_sync_import_end`
* Both events should have two parameters (`wordpress`, `WordPress`)
* Other types of imports that trigger the `import_done` or `import_end` actions should also sync `jetpack_sync_import_done`. Examples: (Blogger, LiveJournal, RSS). Each should have two parameters, and as long as the importer is in `known_importers`, those parameters should reflect those values.  These other importers, however, don't currently trigger `import_start`, so `jetpack_sync_import_start` won't be triggered.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->

No changelog entry needed

Commit:
Sync: improve importer detection and sync start action

jblz commented on sync/class.jetpack-sync-module-import.php:
```diff
@@ -1,87 +1,97 @@
+	private static $import_sync_action_map = array(
+		'import_start' => 'jetpack_sync_import_start',
+		'import_done'  => 'jetpack_sync_import_end',
+		'import_end'   => 'jetpack_sync_import_end',
```
It may be worth it to comment as to why `import_done` & `import_end` both map to `jetpack_sync_import_end`

jonathansadowski replied:
Thanks for the suggestion. I just pushed a commit that adds a comment explaining why they map to the same sync action.

jblz commented on sync/class.jetpack-sync-module-import.php:
```diff
@@ -1,87 +1,97 @@
-		// We already ran an send the import
-		if ( $this->import_end ) {
+		// Only sync each action once per import
+		if ( array_key_exists( $sync_action, $this->synced_actions ) && $this->synced_actions[ $sync_action ] ) {
```
If some request triggers both `import_done` & `import_end` it looks like this would limit it to a single `jetpack_sync_import_end`, correct?  That seems like the desired behavior here, so 👍 if so :)

Is there any instance where these would be triggered in separate request cycles for the same import?  If so, is it acceptable that we'd have repeat sync events?  Probably not a huge issue if so.

jonathansadowski replied:
That is correct. The replaced code worked similarly (though, it only handled a single action mapping, ensuring `jetpac_sync_import_end` was only triggered once). The update 

While I suppose that it's possible there is an importer that would trigger multiple `import_done` or `import_end` actions outside of a single request, none of the primary importers that we're syncing currently do. If we find this to be the case, we can certainly revisit -- but since it's something that importers _shouldn't_ do, until and unless we see it in practice, I think it's acceptable.

jblz reviewed (commented):
This looks good and is working well for me, but will hold off marking my approval until the "synced wpcom file" issue is resolved. I left a couple of questions in the meantime.

Commit:
Add comment clarifying why import_done and import_end map to the same event

jonathansadowski commented:
Thanks for the review @jblz. It looks like the process of pushing my previous commit caused the block by p9dueE-wa-p2 to no longer apply. I'll review D25918-code.

Commit:
Move get_calling_importer_class out of Jetpack_Sync_Functions

jblz commented on sync/class.jetpack-sync-module-import.php:
```diff
@@ -2,86 +2,148 @@
+		$do_action_pos = -1;
+		for ( $i = 0; $i < count( $backtrace ); $i++ ) {
+			// Find the location in the stack of the calling action
+			if ( preg_match( "/^do_action\\(\'([^\']+)/", $backtrace[ $i ], $matches ) ) {
```
Is it worth guarding against implementation changes in `wp_debug_backtrace_summary`, etc. by allowing for optional whitespace around the argument in `do_action`?

Something like this:

```suggestion
			if ( preg_match( "/^do_action\s*\\(\s*\'([^\']+\s*)/", $backtrace[ $i ], $matches ) ) {
```

jonathansadowski replied:
Thanks, great idea -- I just committed your suggestion!

jblz committed:
Be more forgiving with `wp_debug_backtrace_summary`

Allow for `wp_debug_backtrace_summary` to contain whitespace to guard against future changes.

Co-Authored-By: jonathansadowski <jonathansadowski@users.noreply.github.com>

jblz reviewed (dismissed):
This is working well! Accepting on behalf of Samus Team.

kraftbj commented:
Asking for a set of Poseidon eyes. The changes look reasonable to me, though there are some new PHPCS issues added. I'd prefer that the changed lines don't introduce new issues, but that isn't a blocker for this file yet.

Commit:
Fix PHPCS issues

jonathansadowski commented:
Thanks for adding @Automattic/poseidon to the review @kraftbj!

I've updated the PR to address the PHPCS issues.

Commit:
PHPCS fix: missing full-stop in comment

lezama commented on sync/class.jetpack-sync-module-import.php:
```diff
@@ -2,86 +2,153 @@
+
+		$do_action_pos = -1;
+		$backtrace_len = count( $backtrace );
+		for ( $i = 0; $i < $backtrace_len; $i++ ) {
```
have you seen https://github.com/Automattic/jetpack/blob/dd9c8a374f4a136e4499853becb124b4270561d3/class.jetpack.php#L6965 maybe we could extend that one or add a new method to Jetpack.class

jonathansadowski replied:
Thanks for the suggestion @lezama. I wasn't aware of that method prior to creating `get_calling_importer_class`. I explored creating something similar in `Jetpack.class`, but there are a few issues with such an approach:

* The existing method simply returns _if_ any of an array of names is a function in the backtrace. It'd be pretty trivial to extend that to support any column of the backtrace (not just `function`, but also `class`), but we care about _which_ class is found in the backtrace.
* We currently care not only about known importers, but any class that extends `WP_Importer`. We could of course add logic similar to the logic we have in `get_calling_importer_class` that checks for subclasses, but that would undo many of the optimizations currently in place in `is_function_in_backtrace`
* All of these extra needs are really very specific to our implementation of getting the calling class in Sync for importers. Until we have the need to do something similar outside of Jetpack Sync, I'm hesitant to write something more generalized without a better idea of how it'd be used.

enejb commented on sync/class.jetpack-sync-module-import.php:
```diff
@@ -2,86 +2,153 @@
 
-		/** This filter is already documented in sync/class.jetpack-sync-module-posts.php */
-		do_action( 'jetpack_sync_import_end', $importer, $importer_name );
+		do_action( $sync_action, $importer, $importer_name );
```
It would be really great to leave a comment here. So when someone searched for `jetpack_sync_import_start` or `jetpack_sync_import_end` we also get this line.

jonathansadowski replied:
Great suggestion. Comment added!

lezama commented on sync/class.jetpack-sync-module-import.php:
```diff
@@ -2,86 +2,153 @@
+	 *
+	 * @var array
+	 */
+	private static $known_importers = array(
```
we could use [Sync Settings](https://github.com/Automattic/jetpack/blob/master/sync/class.jetpack-sync-settings.php) here, the good thing about it is that they are easy to update remotely or via cli if one day we want to add a new importer to the list without waiting for a release.

creativecoder replied:
I've updated the PR to use a default sync setting for `known_importers`, using the conventions I found in `Jetpack_Sync_Settings` and `Jetpack_Sync_Defaults`

kraftbj commented:
Marking as needing @jonathansadowski's responses. Note that Monday EOD is code freeze for 7.3, so love to get this wrapped up and in the next couple of business days.

creativecoder committed:
Uses sync settings for known importers

Commit:
Because we're dealing with backtrace, ensure class_exists and class_parents don't autoload

Commit:
Use debug_backtrace instead of wp_debug_backtrace_summary

Commit:
Add comment about sync_action

creativecoder committed:
Adds tests for importer sync updates

creativecoder committed:
Merge branch 'update/import-sync-detection' of github.com:Automattic/jetpack into update/import-sync-detection

creativecoder commented:
@kraftbj @lezama I believe we've addressed all the outstanding comments if it's still possible to get this in for 7.3 🤞.

kraftbj commented on sync/class.jetpack-sync-module-import.php:
```diff
@@ -1,87 +1,143 @@
-		/** This filter is already documented in sync/class.jetpack-sync-module-posts.php */
-		do_action( 'jetpack_sync_import_end', $importer, $importer_name );
+		// $sync_action should be either 'jetpack_sync_import_start' or 'jetpack_sync_import_end' as mapped via $import_sync_action_map.
+		do_action( $sync_action, $importer, $importer_name );
```
I think this is going to trip up our developer.jetpack.com parser, or at least not be able to parse it.

@jeherve Do you have ideas of how to document this? We could add the two `do_action` calls in another file that is never called for the sake of documentation, linking back to here as the actual place it fires? I'm okay adding that after merging to not block this so we can get to freeze.

jonathansadowski replied:
Since there are only 2 possible `$sync_action`s, we could also just use a conditional or a switch statement here...

```
switch( $sync_action ) {
	case 'jetpack_sync_import_start':
		do_action( 'jetpack_sync_import_start' ... );
		break;
	...
}
```

jeherve replied:
Yeah, let's go with a conditional. It's not super pretty, but it should make the parser happy and may actually end up being easier to get back to later, when looking for specific actions for example.

kraftbj replied:
Thank you both. Let's do the switch and add inline documentation above each do_action. @jonathansadowski do you have bandwidth to knock that out now-ish? If not, I can get to it in a short bit.

jonathansadowski replied:
Yep, grabbing it right now. Thanks @jeherve & @kraftbj.

Commit:
Move sync actions to switch statement; document
